### ISSUE TITLE
Design Question Classifier (Learned vs Rule-Based)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#37]
## Overview

Decide on approach for classifying question types (Kiu/Kio/Kie/etc) to select appropriate answer templates. This is **required** for template-based answer generation (Issue #36).

## Problem

Issue #36 (Answer Generation) describes template approach:
1. **Classify question type** → Select template
2. Fill template slots → Generate answer AST
3. Deparse → Final response

**But we haven't designed the question classifier!**

## Question Types (Esperanto)

Need to classify ~8 question types:

| Question Word | Type | Example | Answer Template |
|---------------|------|---------|-----------------|
| **Kiu** | Identity (who) | "Kiu kreis Esperanton?" | "{X} kreis {Y}" |
| **Kio** | Definition (what) | "Kio estas Esperanto?" | "{X} estas {Y}" |
| **Kie** | Location (where) | "Kie oni parolas Esperanton?" | "Oni parolas {X} en {Y}" |
| **Kiam** | Time (when) | "Kiam kreiĝis Esperanto?" | "{X} kreiĝis en {Y}" |
| **Kial** | Reason (why) | "Kial Esperanto estas facila?" | "{X} estas {Y} ĉar {Z}" |
| **Kiel** | Manner (how) | "Kiel oni lernas Esperanton?" | "Oni lernas {X} per {Y}" |
| **Kiom** | Quantity (how many) | "Kiom da homoj parolas Esperanton?" | "{X} homoj parolas {Y}" |
| **Ĉu** | Yes/no | "Ĉu Esperanto estas facila?" | "Jes/Ne, {X} estas {Y}" |

## Option A: Rule-Based (AST Pattern Matching)

**Approach**: Extract question word from query AST deterministically

```python
class QuestionClassifier:
    """Rule-based question type classification from AST."""
    
    QUESTION_WORDS = {
        "kiu": "identity",
        "kio": "definition",
        "kie": "location",
        "kiam": "time",
        "kial": "reason",
        "kiel": "manner",
        "kiom": "quantity",
        "ĉu": "yes_no"
    }
    
    def classify(self, query_ast: Dict) -> str:
        """
        Extract question word from AST structure.
        
        Strategy:
        1. Find interrogative pronoun in AST
        2. Map to question type
        3. Return type string
        """
        # Search AST for question word
        for node in ast_traverse(query_ast):
            if node.get("vortspeco") == "demanda_pronomo":
                radiko = node.get("radiko")
                return self.QUESTION_WORDS.get(radiko, "unknown")
        
        # Fallback: check for ĉu at sentence start
        if self.starts_with_cu(query_ast):
            return "yes_no"
        
        return "unknown"  # Default to extractive answer
```

**Advantages**:
- ✅ 0 parameters (pure symbolic)
- ✅ 100% accuracy on well-formed questions
- ✅ Fast (no inference)
- ✅ Explainable
- ✅ Aligned with AST-first philosophy

**Disadvantages**:
- ⚠️ Requires well-parsed AST
- ⚠️ May miss edge cases
- ⚠️ Cannot handle ambiguous questions

**Implementation**: 2-3 days

---

## Option B: Learned Classifier (Neural)

**Approach**: Train classifier on query representations

```python
class LearnedQuestionClassifier(nn.Module):
    """Neural question type classifier."""
    
    def __init__(self, input_dim: int = 256, num_types: int = 8):
        super().__init__()
        
        # Classifier head on Tree-LSTM representation
        self.classifier = nn.Sequential(
            nn.Linear(input_dim, 128),
            nn.ReLU(),
            nn.Dropout(0.1),
            nn.Linear(128, num_types)
        )
    
    def forward(self, query_repr: Tensor) -> Tensor:
        """
        query_repr: [256d] from Tree-LSTM encoder
        Returns: [8] logits over question types
        """
        return self.classifier(query_repr)
```

**Parameters**: ~33K (tiny!)

**Training Data**: 10K-50K labeled questions
```python
{
    "question": "Kiu kreis Esperanton?",
    "query_ast": {...},
    "question_type": "identity"  # Label
}
```

**Training Objective**: Multi-class cross-entropy

**Advantages**:
- ✅ Robust to parse errors
- ✅ Can learn implicit patterns
- ✅ Handles ambiguous cases
- ✅ Very small model (33K params negligible)

**Disadvantages**:
- ⚠️ Requires training data labeling
- ⚠️ Training time (~1-2 days)
- ⚠️ Adds learned component

**Implementation**: 1 week (including data labeling)

---

## Option C: Hybrid Approach

**Approach**: Rule-based with learned fallback

```python
class HybridQuestionClassifier:
    def __init__(self, learned_classifier=None):
        self.rule_based = RuleBasedClassifier()
        self.learned = learned_classifier  # Optional
    
    def classify(self, query_ast: Dict, query_repr: Tensor = None) -> str:
        # Try rule-based first
        q_type = self.rule_based.classify(query_ast)
        
        if q_type != "unknown":
            return q_type
        
        # Fallback to learned if available
        if self.learned and query_repr is not None:
            return self.learned(query_repr)
        
        return "extractive"  # Final fallback
```

**Advantages**:
- ✅ Best of both worlds
- ✅ Explainable for simple cases
- ✅ Handles edge cases

**Disadvantages**:
- ⚠️ More complex

---

## Recommendation

**For M1**: **Option A** (Rule-Based)

**Rationale**:
1. **Esperanto regularity**: Question words are always explicit in AST
2. **AST-first philosophy**: Maximize symbolic processing
3. **Speed**: Can implement in 2-3 days vs 1 week
4. **Simplicity**: No training data needed
5. **Coverage**: 8 question types cover 95%+ of factual questions

**For M2**: Consider **Option C** (Hybrid) if rule-based insufficient

---

### Week 1: Design & Implementation

**Day 1-2**: Design
- [ ] Document AST patterns for each question type
- [ ] Design AST traversal algorithm
- [ ] Define fallback strategy for unknown types

**Day 3**: Implementation
- [ ] Implement `QuestionClassifier` class
- [ ] Add AST traversal utilities
- [ ] Handle edge cases

**Day 4**: Testing
- [ ] Create test cases for 8 question types
- [ ] Test on Q&A dataset (Issue #24)
- [ ] Measure coverage (% questions classified)

**Day 5**: Integration
- [ ] Integrate with Issue #36 (Answer Generation)
- [ ] Update template selector
- [ ] End-to-end testing

---

## Success Criteria

✅ **Rule-Based Approach**:
- Correctly classifies 95%+ of well-formed questions
- 0 learned parameters
- <1ms inference time
- 100% explainable

✅ **Learned Approach** (if chosen):
- 97%+ accuracy on test set
- <33K learned parameters
- Generalizes to unseen questions

---

## Dependencies

**Blocks**:
- Issue #36: Answer Generation (cannot select templates without this)
- Issue #25: Reasoning Core Training (needs end-to-end pipeline)

**Blocked by**:
- None (can start immediately)

**Related**:
- Issue #24: Q&A Data Generation (provides test questions)

---

## Priority

**P0** - Critical Path

This is a **prerequisite** for Issue #36, which is required for M1 completion.

**Timeline**:
- Rule-based: 3-5 days
- Learned: 1 week
- Hybrid: 1-2 weeks

---

## Open Questions

1. **Should we handle multi-word questions?**
   - "Kiu kaj kio" (who and what)
   - "Kiam kaj kie" (when and where)
   - Recommendation: Classify as first question word for M1

2. **How to handle indirect questions?**
   - "Mi volas scii, kiu kreis Esperanton"
   - Recommendation: Parser should normalize to direct question

3. **What about rhetorical questions?**
   - "Kiu ne ŝatas Esperanton?" (Who doesn't like Esperanto?)
   - Recommendation: Classify normally, let reasoning core handle

4. **Fallback for failed classification?**
   - Return "extractive" → use RAG baseline
   - Return "unknown" → ask clarification
   - Recommendation: Extractive fallback for M1

---

## Related Issues

- Issue #36: Implement Answer Generation (requires this)
- Issue #24: Generate Q&A Training Data (provides test cases)
- Issue #35: RAG Integration (extractive fallback)

**Comment by marctjones** [comment]
Phase 6 cleanup - see CURRENT_PLAN.md for active work
